### PR TITLE
Refactor controllers and contexts error handling

### PR DESCRIPTION
**Pull request by fabriziosestito**
Refactor web error handling:
- Make controllers agnostic of domain error
- Make fallback controller aware of domain error
- Fix error handling in certain contexts, improve logging
- Add changeset rendering in ErrorView
- use a tagged tuple to return validation errors

### CODE REVIEW

File: lib/trento/application/usecases/clusters/clusters.ex [hunk 1]
```diff
@@ -71,7 +82,9 @@ defmodule Trento.Clusters do
           :ok
 
         {:error, reason} ->
-          Logger.error("Failed to request checks execution for cluster #{cluster_id}: #{reason}")
+          Logger.error(
```
Comment (arbulu89) [inline lib/trento/application/usecases/clusters/clusters.ex:85]: Not something you should do in this PR, but we should reduce the errors in this `each` now and return error if any of them fails.
Or this return the error if fails?

File: lib/trento/application/usecases/hosts/heartbeats.ex [hunk 2]
```diff
@@ -28,15 +28,29 @@ defmodule Trento.Heartbeats do
...
+
+    case result do
+      {:ok, _} ->
+        :ok
+
+      {:error, :command, :host_not_registered, _} ->
+        {:error, :not_found}
+
+      error ->
+        Logger.error("Error while updating heartbeat for agent #{agent_id}", error: inspect(error))
```
Comment (arbulu89) [inline lib/trento/application/usecases/hosts/heartbeats.ex:50]: Nothing wrong here, but i see that we use the error metadata here, but in every other place we include the error in the main message text

File: lib/trento_web/controllers/fallback_controller.ex [hunk 3]
```diff
@@ -3,31 +3,60 @@ defmodule TrentoWeb.FallbackController do
...
     conn
     |> put_status(:not_found)
     |> put_view(ErrorView)
-    |> render(:"404", detail: detail)
+    |> render(:"404")
   end
 
-  def call(conn, {:error, {:unauthorized, detail}}) do
+  def call(conn, {:error, reason})
+      when reason in [:host_not_registered, :cluster_not_registered, :sap_system_not_registered] do
```
Comment (arbulu89) [inline lib/trento_web/controllers/fallback_controller.ex:28]: I hope this list doesn't convert on a root for bugs hehe

**Review by arbulu89 — APPROVED** [review]
Cool!

As you have added/updated error logging, did you check if we are capturing them in the tests? Just to have a clean tests output, and not filled with "expected" error messages

File: lib/trento/application/usecases/hosts/heartbeats.ex (hunk 2 above)
Comment (fabriziosestito) [reply lib/trento/application/usecases/hosts/heartbeats.ex:50]: ah yes, we did not decided yet where and when to use metadata so I will add the error to the message and defer this decision

File: lib/trento/application/usecases/clusters/clusters.ex (hunk 1 above)
Comment (fabriziosestito) [reply lib/trento/application/usecases/clusters/clusters.ex:85]: this is used by the job scheduler in a best-effort manner, this is why we just use the each loop and we return nothing.
If something went wrong while scheduling the execution we log it

**Review by nelsonkopliku — APPROVED** [review]
Nice job, thanks!

**Pull request merged**